### PR TITLE
CO-11932 : Enable aws cookbook with dynamic secret

### PR DESCRIPTION
**Pull request by Ramesh7**
## JIRA

* [Main JIRA ticket](https://coupadev.atlassian.net/browse/CO-11932)

## Summary of issue

Enable aws cookbook resources for storage to use federation token.

## Summary of change

Enable aws cookbook resources for storage to use federation token.

## Testing approach

- [ ] Serverspec test (mandatory)
- [x] Manual test
- [ ] No test (please specify why)

### CODE REVIEW

File: libraries/ec2.rb [hunk 1]
```diff
@@ -20,30 +20,26 @@
 module Opscode
   module Aws
     module Ec2
-      def find_snapshot_id(volume_id = '', find_most_recent = false)
+      def find_snapshot_id(volume_id = '', find_most_recent = false, options = {})
```
Comment (Ramesh7) [inline libraries/ec2.rb:23]: Adding filter to search snapshot via filter options.

File: libraries/ec2.rb [hunk 2]
```diff
@@ -20,30 +20,26 @@
...
-                   end
-        response.each do |page|
-          page.snapshots.each do |snapshot|
-            if snapshot[:volume_id] == volume_id && snapshot[:state] == 'completed'
-              snapshot_id = snapshot[:snapshot_id]
-            end
-          end
-        end
-        raise 'Cannot find snapshot id!' unless snapshot_id
+                     ec2.describe_snapshots(filters).sort { |a, b| b[:start_time] <=> a[:start_time] }
```
Comment (Ramesh7) [inline libraries/ec2.rb:31]: should not raise if existing snapshot not available. Used while launching new deployments.

File: libraries/ec2.rb [hunk 3]
```diff
@@ -20,30 +20,26 @@
...
+                   end.first.snapshots.first
+        snapshot_id = response[:snapshot_id] if response
         Chef::Log.debug("Snapshot ID is #{snapshot_id}")
         snapshot_id
       end
 
       def ec2
         begin
-          require 'aws-sdk'
+          require 'aws-sdk-ec2'
```
Comment (Ramesh7) [inline libraries/ec2.rb:40]: requiring specific library.

File: providers/ebs_volume.rb [hunk 4]
```diff
@@ -83,32 +92,113 @@ def whyrun_supported?
...
     end
   end
-  if old_snapshots.length > new_resource.snapshots_to_keep
-    old_snapshots[new_resource.snapshots_to_keep, old_snapshots.length].each do |die|
-      converge_by("delete snapshot with id: #{die[:snapshot_id]}") do
-        Chef::Log.info "Deleting old snapshot #{die[:snapshot_id]}"
-        ec2.delete_snapshot(snapshot_id: die[:snapshot_id])
-      end
+
+  (1..new_resource.snapshots_keep_yearlies).each do |y|
```
Comment (Ramesh7) [inline providers/ebs_volume.rb:149]: Custom cleanup lifecycle for snapshots.

File: providers/ebs_volume.rb [hunk 5]
```diff
@@ -83,32 +92,113 @@ def whyrun_supported?
...
-    Chef::Log.info("Created snapshot of #{vol[:volume_id]} as #{snapshot[:volume_id]}")
+    Chef::Log.info("Created snapshot of #{vol[:volume_id]} as #{snapshot[:snapshot_id]}")
+
+    node.normal['aws']['ebs_volume'][new_resource.name]['snapshots'] = (node['aws']['ebs_volume'][new_resource.name]['snapshots'] || []) + [snapshot[:snapshot_id]]
+    node.save unless Chef::Config[:solo]
+
+    tags = Hash[new_resource.snapshot_filters.select{|x| x.start_with?("tag:")}.map{|k, v| [k[4..-1], v]}]
+    tags.merge!('timestamp' => Time.now.to_i.to_s)
+    volume_name = new_resource.name
+    aws_resource_tag "Tagging the latest snapshot of volume #{vol[:volume_id]}" do
```
Comment (Ramesh7) [inline providers/ebs_volume.rb:125]: Tagging snapshot with lineage & timestamp.

File: recipes/default.rb [hunk 6]
```diff
@@ -17,10 +17,5 @@
 # limitations under the License.
 #
 
-chef_gem 'aws-sdk' do
```
Comment (Ramesh7) [inline recipes/default.rb:20]: Removing aws-sdk installation as we already installing cloud_resource which has runtime dependencies on aws-sdk-*

**Review by pallav-jakhotiya — APPROVED** [review]

**Review by kavita-borkar — APPROVED** [review]

**Pull request merged**